File: Intervals/p56_merge_intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key= lambda x: x[0])
        print(intervals)
        res=[]

        i=0
        while i< (len(intervals)):

            lowBound=intervals[i][0]
            highBound=intervals[i][1]
            while i+1<len(intervals) and highBound>=intervals[i+1][0]:
                highBound=max(highBound, intervals[i+1][1])
                i+=1
            i+=1
            
            res.append([lowBound,highBound])

        
        return res


   
            
#time complexity O(NLogN), from the sorting
#space complexity O(N)


            

File: Intervals/p56_merge_intervals.cpp
class Solution {
public:
  vector<vector<int>> merge(vector<vector<int>> &intervals) {
    sort(intervals.begin(), intervals.end());
    int i = 0;
    vector<vector<int>> res;
    while (i < intervals.size()) {
      int start = intervals[i][0];
      int end = intervals[i][1];
      while (i + 1 < intervals.size() && end >= intervals[i + 1][0]) {
        start = std::min(start, intervals[i + 1][0]);
        end = std::max(end, intervals[i + 1][1]);
        i++;
      }
      res.push_back({start, end});
      i++;
    }
    return res;
  }
};
// time complexity O(NLogN)
// space complexity O(N)